### ISSUE TITLE
Podstatus becomes MatchNodeSelector after restart kubelet

### ISSUE THREAD

**Issue by ailusazh** [kubernetes/kubernetes#52902]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
 /kind bug
/sig node

**What happened**:
After restart kubelet, kubelet receives infomation about pods and the node from etcd, in the situation that first receives pods infomaion and start admit pods and then receives node infomation, then in the process of admit pods some will use the initial node info, and if the labels in initial node info different from the node info in etcd, then the admit for the pod will fail and kubelet will set the pod status to MatchNodeSelector.
The frequency of this situation is very low.

**What you expected to happen**:
Should we wait until we get the node infomation from etcd and then start to admit pods?

**How to reproduce it (as minimally and precisely as possible)**:
Any suggestions?

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration**:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by lichuqiang** [comment]
We met the same issue before, but it seems not suggested to label nodes via API(Remembered seeing this conclusion in a issue, but could not find it right now), To label nodes through local configuration with kubelet flag "--node-labels" could avoid the sync matter.
Anyway, we added the check for sync in kubelet in our internal version.

**Comment by ailusazh** [comment]
@lichuqiang we have tried to keep labels in etcd and configuration the same, but there will always be cases of Inconsistent due to negligence, and label node via API is much more convenient. I intend to wait in the admit process until get node info from etcd instead of use initial node info from local configuration. Like this:
```
func (w *predicateAdmitHandler) Admit(attrs *PodAdmitAttributes) PodAdmitResult {
	var node *v1.Node
	for {
		var err error
		node, err = w.getNodeFunc()
		if err == nil {
			break
		}
		time.Sleep(getNodeInternal)
		glog.Infof("Kubelet cannot get node info.")
	}
        ...
```
If getNodeFunc() can't success always, the sync pod process will block here, but this will rarely happen. What about your opinion？

**Comment by lichuqiang** [comment]
@ailusazh Well, the modification may have effects on the behevior of static pods that don't rely on data in etcd indeed.

**Comment by ailusazh** [comment]
Thanks for your reminder. Yes, this does affect static pods, however considered that we don't use static pods, I will add a switch to use this logic.

**Comment by chestack** [comment]
meet same issue，k8s v1.9.8

Except @ailusazh has commented that kubelet should retry to getNode(), It seems that kubelet does not re sync pods after MatchNodeSelector failed events. Found following logs:
` Reason:"MatchNodeSelector", Message:"Predicate MatchNodeSelector failed", Source:v1.EventSource{Component:"kubelet", Host:"node-1"}, FirstTimestamp:v1.Time{Time:time.Time{wall:0xbee665c8e6df88de, ext:224008353822, loc:(*time.Location)(0xaa9ee60)}}, LastTimestamp:v1.Time{Time:time.Time{wall:0xbee665c8e6df88de, ext:224008353822, loc:(*time.Location)(0xaa9ee60)}}, Count:1, Type:"Warning", EventTime:v1.MicroTime{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, Series:(*v1.EventSeries)(nil), Action:"", Related:(*v1.ObjectReference)(nil), ReportingController:"", ReportingInstance:""}': 'etcdserver: request timed out' (will not retry!)`

'Timeout: request did not complete within allowed duration' (will not retry!)
'etcdserver: request timed out' (will not retry!)

**Comment by vladwa** [comment]
I have seen this happening when the "nodeName" value is different from the "nodeSelector".

**Comment by hustcat** [comment]
+1 for this issue.

**Comment by Bevisy** [comment]
maybe fixed by #94087

**Comment by ehashman** [comment]
@Bevisy I agree, I think this is fixed by that patch, especially given the comments around `GetNode()`.
